tests/Unit/Lti/Factory/Lti1p3RequestFactoryTest.php

```diff
@@ -0,0 +1,179 @@
+                'ltiParams' => [],
+            ],
+            $this->subject->create($assignment)->jsonSerialize()
+        );
+
+        Carbon::setTestNow();
+    }
+
+    public function provideValidLineItemAssignmentCombinations(): array
+    {
+        return [
+            'withMaxAttemptsSpecifiedAndAvailableAttemptsAndReadyStatus' => [
+                'maxAttempts' => 2,
+                'attemptsCount' => 1,
+                'assignmentState' => Assignment::STATE_READY,
+            ],
+            'withMaxAttemptsSpecifiedAndAvailableAttemptsAndStartedStatus' => [
+                'maxAttempts' => 2,
+                'attemptsCount' => 1,
+                'assignmentState' => Assignment::STATE_STARTED,
+            ],
+            'withMaxAttemptsSpecifiedAndLastAttemptStartedAndStartedStatus' => [
+                'maxAttempts' => 3,
+                'attemptsCount' => 3,
+                'assignmentState' => Assignment::STATE_STARTED,
+            ],
+            'withNoMaxAttemptsSpecifiedAndNoAttemptsTakenAndReadyStatus' => [
+                'maxAttempts' => 0,
+                'attemptsCount' => 0,
+                'assignmentState' => Assignment::STATE_READY,
+            ],
+            'withNoMaxAttemptsSpecifiedAndSomeAttemptsTakenAndReadyStatus' => [
+                'maxAttempts' => 0,
+                'attemptsCount' => 4,
+                'assignmentState' => Assignment::STATE_READY,
+            ],
+            'withNoMaxAttemptsSpecifiedAndSomeAttemptsTakenAndStartedStatus' => [
+                'maxAttempts' => 0,
+                'attemptsCount' => 4,
+                'assignmentState' => Assignment::STATE_STARTED,
+            ],
+        ];
+    }
+
+    public function testShouldThrowRegistrationNotFoundException(): void
+    {
+        $this->expectException(RegistrationNotFoundException::class);
+        $this->expectExceptionMessage('Registration registrationId not found.');
+
+        $assignment = $this->createAssignmentMock(5, 1, Assignment::STATE_READY);
+
+        $this->subject->create($assignment);
+    }
+
+    private function createAssignmentMock(int $maxAttempts, int $attemptsCount, string $assignmentStatus): Assignment
+    {
+        $lineItem = (new LineItem())
+            ->setMaxAttempts($maxAttempts)
+            ->setUri('http://test-delivery-uri.html')
+            ->setSlug('slug')
+            ->setLabel('label');
+
+        $user = (new User())
+            ->setUsername('testUsername')
+            ->setGroupId('groupId');
+
+        $assignment = $this->createPartialMock(Assignment::class, ['getId']);
+
+        $assignment
+            ->method('getId')
+            ->willReturn(5);
+
+        $assignment
+            ->setLineItem($lineItem)
+            ->setUser($user)
+            ->setAttemptsCount($attemptsCount)
+            ->setState($assignmentStatus);
+
+        return $assignment;
+    }
+
+    private function createRegistrationMock(): void
```

Comment:
This method does not really add value because, also it does more than what it says. You can remove and set mock expectations in your test case

Reply:
I use it on bot tests